test/clj/rems/api/test_applications.clj

```diff
@@ -724,15 +747,169 @@
-      (let [response (-> (upload-request filecontent)
-                         (authenticate api-key user-id)
-                         handler)]
         (is (response-is-forbidden? response))))))
 
+(deftest test-application-comment-attachments
+  (let [api-key "42"
+        applicant-id "alice"
+        handler-id "developer"
+        reviewer-id "carl"
+        file #(assoc filecontent :filename %)
+        workflow-id (test-data/create-workflow! {:type :workflow/master
+                                                 :handlers [handler-id]})
+        cat-item-id (test-data/create-catalogue-item! {:workflow-id workflow-id})
+        application-id (test-data/create-application! {:catalogue-item-ids [cat-item-id]
+                                                       :actor applicant-id})
+        add-attachment #(-> (request :post (str "/api/applications/add-attachment?application-id=" application-id))
+                            (authenticate api-key %1)
+                            (assoc :params {"file" %2})
+                            (assoc :multipart-params {"file" %2})
+                            handler
+                            read-ok-body
+                            :id)]
+    (testing "submit"
+      (is (= {:success true} (send-command applicant-id
+                                           {:type :application.command/submit
+                                            :application-id application-id}))))
+    (testing "unrelated user can't upload attachment"
+      (is (response-is-forbidden? (-> (request :post (str "/api/applications/add-attachment?application-id=" application-id))
+                                      (authenticate api-key reviewer-id)
+                                      (assoc :params {"file" filecontent})
+                                      (assoc :multipart-params {"file" filecontent})
+                                      handler))))
+    (testing "invite reviewer"
+      (is (= {:success true} (send-command handler-id
+                                           {:type :application.command/request-review
+                                            :application-id application-id
+                                            :reviewers [reviewer-id]
+                                            :comment "please"}))))
+
+    (testing "handler uploads an attachment"
+      (let [attachment-id (add-attachment handler-id (file "handler-remark.txt"))]
+        (is (number? attachment-id))
+        (testing "and attaches it to a public remark"
+          (is (= {:success true} (send-command handler-id
+                                               {:type :application.command/remark
+                                                :application-id application-id
+                                                :comment "see attachment"
+                                                :public true
+                                                :attachments [attachment-id]}))))))
```

Comment:
needs a test for not being able to remove the attachment after it is in use

Reply:
no, there's no api for removing attachments (I was thinking about license attachments, which can be removed if they're not in use)